packages/avatar-creator/src/components/MmlOverlayImport.tsx

```diff
@@ -0,0 +1,225 @@
+/**
+ * @license
+ * Copyright Improbable MV Limited.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://github.com/msquared-io/avatar-creator/blob/main/LICENSE
+ */
+
+import hljs from "highlight.js";
+import xml from "highlight.js/lib/languages/xml";
+import * as React from "react";
+import { MouseEvent, useEffect, useRef } from "react";
+
+import { CatalogueBodyType } from "../CatalogueData";
+import { AvatarLoader } from "../scripts/avatar-loader";
+import styles from "./MmlOverlayImport.module.css";
+
+hljs.registerLanguage("xml", xml);
+
+const keyReplace = {
+  topSecondary: "top:secondary",
+  bottomSecondary: "bottom:secondary",
+};
+
+export default function MmlOverlayImport({
+  setActive,
+  avatarLoader,
+}: {
+  setActive: (value: string) => void;
+  avatarLoader: AvatarLoader;
+}) {
+  const ref = useRef<HTMLDivElement>(null);
+  const codeRef = useRef<HTMLPreElement>(null);
+  let overlayDown = false;
+  let focusCode = false;
+
+  const selectCode = () => {
+    if (!codeRef.current) return false;
+    const range = document.createRange();
+    range.selectNodeContents(codeRef.current);
+    const selection = window.getSelection();
+    if (!selection) return false;
+    selection.removeAllRanges();
+    selection.addRange(range);
+    return true;
+  };
+
+  useEffect(() => {
+    selectCode();
+
+    navigator.clipboard
+      .readText()
+      .then((clipText) => {
+        if (codeRef.current) codeRef.current.innerText = clipText;
+      })
+      .catch(() => {});
+
+    const evtWindowMouseDown = (evt: globalThis.MouseEvent) => {
+      overlayDown = evt.target === ref.current;
+    };
+    const evtWindowMouseUp = (evt: globalThis.MouseEvent) => {
+      if (overlayDown && evt.target === ref.current) setActive("");
+      overlayDown = false;
+    };
+
+    window.addEventListener("mousedown", evtWindowMouseDown);
+    window.addEventListener("mouseup", evtWindowMouseUp);
+
+    return () => {
+      window.removeEventListener("mousedown", evtWindowMouseDown);
+      window.removeEventListener("mouseup", evtWindowMouseUp);
+    };
+  }, []);
+
+  const onCodeMouseDown = () => {
+    focusCode = false;
+    if (!codeRef.current) return false;
+    if (codeRef.current === document.activeElement) return;
+    selectCode();
+    focusCode = true;
+  };
+
+  const onCodeMouseUp = (evt: MouseEvent) => {
+    if (focusCode) {
+      focusCode = false;
+      evt.preventDefault();
+    }
+  };
+
+  const onImport = () => {
+    const code = codeRef.current?.textContent ?? "";
+
+    const parser = new DOMParser();
```

Comment:
Can you extract both this and the export equivalent to functions that can share some literals to maybe make the symmetry clearer please? Just conscious that things like the string parsing for `skin` are a difficult to discover the counterpart of.